holding/src/client/mod.rs

```diff
@@ -174,23 +175,74 @@ where
         //.inspect(|(_, body)| debug!("Response body: {:?}", ::std::str::from_utf8(body.as_ref())))
         .and_then(move |(httpstatus, body)| -> Result<T, Error> {
             if !httpstatus.is_success() {
+                // I think we can drop this debug! it is redundant with enrich_error.
                 debug!("failure body: {:#?}", ::std::str::from_utf8(body.as_ref()));
                 let status: Status = serde_json::from_slice(body.as_ref()).map_err(|e| {
                     debug!(
                         "Failed to parse error Status ({}), falling back to HTTP status",
-                        e
+                        enrich_error("error Status", &e, body.as_ref())
                     );
                     HttpStatusError { status: httpstatus }
                 })?;
                 Err(status.into())
             } else {
                 let o = serde_json::from_slice(body.as_ref())
-                    .with_context(|e| format!("Unable to parse response body: {}", e))?;
+                    .with_context(|e| enrich_error("response body", e, body.as_ref()))?;
                 Ok(o)
             }
         })
 }
 
+/// Grabs the 1K preceeding text from the failed document and describes the error.
+///
+/// TODO: handle multi-line JSON, just in case some API server decides to start emitting that.
+fn enrich_error(desc: &str, e: &serde_json::Error, body_ref: &[u8]) -> String {
+    // debug! so that operators running with debug logs get *everything*
+    debug!("Parse failure: {:#?}", body_ref);
+    // Provide a short snippet for errors that may be handled, logged at higher verbosity etc.
+    match e.classify() {
+        serde_json::error::Category::Io | serde_json::error::Category::Eof => {
+            format!("Unable to parse {}: {}", desc, e)
+        }
+        _ => {
```

Comment:
I would rather we kept the _entire_ body for these non-I/O errors, and then reduce it to a line or so of context only when displaying (via `Display`) these types of errors.  (iow: move the following code to the `Error` `Display` implementation.)
In the scheme of things, the entire body is a tiny amount of ram, and I think the potential debugging value will be high.

I think you can do this by creating a new `DeserializeError` class that captures the original string and serde_json error, give it a handy constructor, and then implement expected `Display`, `Fail` traits.

Related: Is premature `Eof` a (permanent) deserialize error or a (transient) I/O issue?  I think it could be either, depending on whether the HTTP response had a content-length header, which is more complex than I want :(

Reply:
Lets merge this to make development easier. And I'll put together what you propose as a separate improvement.

Eof is permanent (we buffer and call today, rather than giving serde_json a readable). This could change in future.

Reply:
> Lets merge this to make development easier.

Boo :(   Sure, it doesn't change any of the signatures, it's just a minor improvement to an information error description.  lgtm.

> Eof is permanent

Eg: What happens if the apiserver segfaults halfway through writing a reply body?  I think this will appear as a premature "clean" EOF on the socket, even though it is (from a high-level pov) a transient error.  Without an HTTP content-length, we don't know it ended early.
*With* an HTTP content-length, we know that we got the full content, and failed to understand it (a local/permanent error).

Reply:
We won't call into serde_json when that happens, because a half written message is invalid at the transport protocol layer (except for the special horrid cases of HTTP/0.9 and HTTP/1.0; so while yes we should surface that cleanly, its in a different error family in practice, because all actual apiservers are running modern (20ish years old) HTTP stacks.

Reply:
Oh right, I keep forgetting 1.1 mandated content-length :P
In that case, `Eof` error should be rolled into all the other non-I/O cases when we do the error cleanup.